Functions/Gherkin.Core.ps1

```diff
@@ -0,0 +1,1304 @@
+﻿if (($PSVersionTable.ContainsKey('PSEdition')) -and ($PSVersionTable.PSEdition -eq 'Core')) {
+    & $SafeCommands['Add-Type'] -Path "${Script:PesterRoot}/lib/Gherkin/core/Gherkin.dll"
+}
+else {
+    & $SafeCommands['Import-Module'] -Name "${Script:PesterRoot}/lib/Gherkin/legacy/Gherkin.dll"
+}
+
+$GherkinStepDefinitions = @{ }
+$GherkinHooks = @{
+    BeforeEachFeature  = @()
+    BeforeEachScenario = @()
+    AfterEachFeature   = @()
+    AfterEachScenario  = @()
+}
+
+function Invoke-GherkinHook {
+    <#
+        .SYNOPSIS
+            Internal function to run the various gherkin hooks
+
+        .PARAMETER Hook
+            The name of the hook to run
+
+        .PARAMETER Name
+            The name of the feature or scenario the hook is being invoked for
+
+        .PARAMETER Tags
+            Tags for filtering hooks
+    #>
+    [CmdletBinding()]
+    param([string]$Hook, [string]$Name, [string[]]$Tags)
+
+    if ($GherkinHooks.${Hook}) {
+        foreach ($GherkinHook in $GherkinHooks.${Hook}) {
+            if ($GherkinHook.Tags -and $Tags) {
+                :tags foreach ($hookTag in $GherkinHook.Tags) {
```

Comment:
What is the purpose of this `:tags` declaration here?

Reply:
As well as in line 40 next to the break statement

Reply:
So, I didn't add this label, it was already there. It's more than likely this code appears 'added' because it moved from its previous location in the file.

However, I can answer this question.

The `$Tags` variable that's being enumerated over in the inner loop contains the tags as specified for the `-Tags` parameter to `Invoke-Gherkin`. Now, for scenarios and features and whatnot, if _any_ tag on the scenario, feature, etc. matches one of the tags specified in the `-Tag` parameter, then that scenario, feature, etc., is selected for execution. (Processing of other filters may eventually further exclude these scenarios/features, but, as far as `-Tags` works, my previous statement is accurate.) So, once an "include" tag matches just one of the tags on the hook, we can stop looking for other matching tags because the hook will be included. So we "execute" the hook script block to register it with Pester Gherkin. And because it's been included and registered, we can skip looking for any further tags--there's no point.

Anyway, the one thing I just noticed with this code is that the `ExcludeTag` tags are not processed! I'm thinking this might be a bug. Any thoughts @Jaykul  or @nohwnd? If this is a bug, then I should look to fix this while I'm in here making changes.